### PR TITLE
Add windows Ipc Client implementation

### LINKED ISSUE #2786: Add windows Ipc Client implementation
### Describe the feature

See https://github.com/paradigmxyz/reth/pull/2785

the `IpcClientBuilder::build` returns A sender and receiver that are used to send and receive ipc messages https://github.com/paradigmxyz/reth/blob/c421aeca7f717f915019288fb0352ffcd0872c91/crates/rpc/ipc/src/client.rs#L102-L102
For windows we need types that use named pipes

## TODO

* add Sender + Receiver types that use named pipes instead https://docs.rs/tokio/latest/tokio/net/windows/named_pipe/index.html
* add windows `IpcClientBuilder::build` variant

See https://github.com/paradigmxyz/reth/blob/c421aeca7f717f915019288fb0352ffcd0872c91/crates/rpc/ipc/src/client.rs#L38-L42

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by AbnerZheng**
Close https://github.com/paradigmxyz/reth/issues/2786.

### CODE REVIEW

**Comment by AbnerZheng** [comment]
@mattsse PTAL when you available and let me know if I'm heading in the right direction. Thanks!

File: crates/rpc/ipc/src/client.rs
Comment (mattsse) [inline crates/rpc/ipc/src/client.rs:91]: this is great,

what I would like to do instead is:
convert client into a module with
unix.rs
win.rs

and duplicate the types then client/mod.rs reexports based on windows or UNIX

**Review by mattsse — CHANGES_REQUESTED** [review]
awesome, one suggestion re structure

**Review by onbjerg — CHANGES_REQUESTED** [review]
thank you for taking this on 👑

File: crates/rpc/ipc/src/client/unix.rs
Comment (AbnerZheng) [reply crates/rpc/ipc/src/client/unix.rs:51]: Done. Sorry about this.

**Comment by AbnerZheng** [comment]
But there raise some issues when I tried to use cross build and test.
```rust
error[E0432]: unresolved import `winapi::shared::winerror`
 --> /Users/abnerzheng/.cargo/registry/src/index.crates.io-6f17d22bba15001f/parity-tokio-ipc-0.9.0/src/win.rs:1:21
  |
1 | use winapi::shared::winerror::{ERROR_PIPE_BUSY, ERROR_SUCCESS};
  |                     ^^^^^^^^ could not find `winerror` in `shared`

```
Looks like `parity-tokio-ipc` can't be compiled in windows anymore and out of maintainence. See https://github.com/paritytech/parity-tokio-ipc/issues/39 and https://github.com/paritytech/parity-tokio-ipc/pull/36

The server side implementation relay on `parity-tokio-ipc`.

**Comment by onbjerg** [comment]
I think we only use this for the `Endpoint` type, we should be able to just roll our own, IIRC we had similar issues in Foundry and ditched this crate? cc @mattsse

**Comment by mattsse** [comment]
yeah this is indeed unmaintained and we should actually switch

Is the cross build issue a problem?
I think we can live with this for now?

**Comment by AbnerZheng** [comment]
> Is the cross build issue a problem?
I think if we try to build ipc crate in windows machine, then it will issue the error. 

> I think we can live with this for now?

Yes, but then we have to guard the 'server' mod with #[cfg(unix)] as well, because it won't compile in windows system.

**Comment by mattsse** [comment]
I think we should transition to:

https://github.com/kotauskas/interprocess

**Comment by AbnerZheng** [comment]
> I think we should transition to:
> 
> 
> 
> https://github.com/kotauskas/interprocess

No problem, I will continue the transition in tomorrow.

**Comment by onbjerg** [comment]
Hey @AbnerZheng! Are you blocked on anything here? Let me know 😊

**Comment by AbnerZheng** [comment]
> Let me know 😊

Would take it another try this weekend and will ping you if there are any problem.

**Comment by AbnerZheng** [comment]
@onbjerg @mattsse, when you have a moment, please take a look. The `cargo-deny` check failed. The reason I used the main branch of https://github.com/kotauskas/interprocess is that its latest release 1.2.2, was released in December 2022. However, it is still actively being developed with numerous new features being added, including the incoming method that streamlines the process, along with various bug fixes. Let me know if there is any problem.

File: crates/rpc/ipc/Cargo.toml
Comment (onbjerg) [inline crates/rpc/ipc/Cargo.toml:32]: why a specific commit over the latest release? we'd like to avoid git deps where possible

File: crates/rpc/ipc/src/client/mod.rs
Comment (onbjerg) [inline crates/rpc/ipc/src/client/mod.rs:87]:
```suggestion
    };
    use crate::server::dummy_endpoint;
    use super::*;
```

File: crates/rpc/ipc/src/client/unix.rs
Comment (onbjerg) [inline crates/rpc/ipc/src/client/unix.rs:16]:
```suggestion
use std::path::Path;
use futures::StreamExt;
use jsonrpsee::core::client::{ReceivedMessage, TransportReceiverT, TransportSenderT};
use tokio::{
    io::AsyncWriteExt,
    net::{
        unix::{OwnedReadHalf, OwnedWriteHalf},
        UnixStream,
    },
};
use tokio_util::codec::FramedRead;
use crate::{client::IpcError, stream_codec::StreamCodec};
```

File: crates/rpc/ipc/src/client/win.rs
Comment (onbjerg) [inline crates/rpc/ipc/src/client/win.rs:74]: i understand why this is here, but we might end up indefinitely waiting if the pipe is busy, so i'd like a more robust solution. perhaps we retry x amount of times?

File: crates/rpc/ipc/src/server/mod.rs
Comment (onbjerg) [inline crates/rpc/ipc/src/server/mod.rs:779]: i'd add back `#[cfg(test)]`

File: crates/rpc/rpc-builder/src/lib.rs
Comment (onbjerg) [inline crates/rpc/rpc-builder/src/lib.rs:1493]: i think we can just derive debug now?

**Review by onbjerg — CHANGES_REQUESTED** [review]
coming along nicely! re the deny check, we should either evaluate using 1.2.2, or asking the maintainer to cut a release, we'd like to avoid using git dependencies

**Comment by AbnerZheng** [comment]
> coming along nicely! re the deny check, we should either evaluate using 1.2.2, or asking the maintainer to cut a release, we'd like to avoid using git dependencies

Using 1.2.2 would required to implement the stream method `incoming`,  tried sending a email to maintainer whether it is possible to cut a new release firstly.

**Comment by onbjerg** [comment]
Blocked by https://github.com/kotauskas/interprocess/issues/48

**Comment by AbnerZheng** [comment]
@onbjerg I have switched to use interprocess v1.2.1, which would unblock this PR.

File: deny.toml
Comment (AbnerZheng) [inline deny.toml:35]: @onbjerg @mattsse I  have to allowed this license because `interprocess` use a library which is under it.

File: crates/rpc/ipc/src/server/mod.rs
Comment (AbnerZheng) [inline crates/rpc/ipc/src/server/mod.rs:166]: jsonrpsee has switch to use FuturesUnordered in PR https://github.com/paritytech/jsonrpsee/pull/1062, which address some issues when upgrading tokio. The PR also removed the need of Incoming, Monitored, which unblock this PR.

File: crates/rpc/ipc/src/server/mod.rs
Comment (AbnerZheng) [reply crates/rpc/ipc/src/server/mod.rs:779]: Done.

File: crates/rpc/ipc/src/client/win.rs
Comment (AbnerZheng) [reply crates/rpc/ipc/src/client/win.rs:74]: Would like to open new PR to address this, since we have introduced interprocess here, we don't need use low level named pipe anymore.

File: crates/rpc/ipc/Cargo.toml
Comment (AbnerZheng) [reply crates/rpc/ipc/Cargo.toml]: Done.

File: deny.toml
Comment (onbjerg) [reply deny.toml:35]: can you add it under `exceptions` like the other `CC0-1.0` deps we have?

**Review by onbjerg — APPROVED** [review]
prelim look this looks fine to me, ptal @mattsse 

please move the license exception into where we have our other CC0 deps

File: deny.toml
Comment (AbnerZheng) [reply deny.toml]: Done.

File: crates/rpc/ipc/src/server/mod.rs
Comment (mattsse) [inline crates/rpc/ipc/src/server/mod.rs:231]: this is unreachable and results in unbounded mem

please check jsonrpsee server loop on main

https://github.com/paritytech/jsonrpsee/blob/master/server/src/server.rs

File: crates/rpc/ipc/src/server/mod.rs
Comment (mattsse) [inline crates/rpc/ipc/src/server/mod.rs:447]: this no longer spawns a connections, rename to process

File: crates/rpc/ipc/src/server/mod.rs
Comment (mattsse) [inline crates/rpc/ipc/src/server/mod.rs:852]: can keep it
```suggestion
        init_test_tracing();
```

**Review by mattsse — CHANGES_REQUESTED** [review]
this is great, ty

but futuresunordered has to go

File: crates/rpc/ipc/src/server/mod.rs
Comment (AbnerZheng) [reply crates/rpc/ipc/src/server/mod.rs:231]: Yeah, but would like merge it first, and fire a new PR later today.

File: crates/rpc/ipc/src/server/mod.rs
Comment (AbnerZheng) [reply crates/rpc/ipc/src/server/mod.rs:447]: Done.

File: crates/rpc/ipc/src/server/mod.rs
Comment (AbnerZheng) [reply crates/rpc/ipc/src/server/mod.rs:852]: Done.

File: crates/rpc/ipc/src/server/mod.rs
Comment (mattsse) [reply crates/rpc/ipc/src/server/mod.rs:231]: okay sg, could you please open an issue for this then?

**Review by mattsse — APPROVED** [review]
lgtm,
needs followup re fut unordered

File: crates/rpc/ipc/src/server/mod.rs
Comment (AbnerZheng) [reply crates/rpc/ipc/src/server/mod.rs:231]: https://github.com/paradigmxyz/reth/issues/7916 
would fire a PR today.

**Pull request merged**
